edward/inferences/ali.py

```diff
@@ -0,0 +1,86 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import six
+import tensorflow as tf
+
+from edward.inferences.gan_inference import GANInference
+from edward.util import get_session
+
+
+class ALI(GANInference):
+  """Adversarially Learned Inference (Dumoulin et al., 2016) or
+  Bidirectional Generative Adversarial Networks (Donahue et al., 2016)
+  for joint learning of generator and inference networks.
+
+  Works for the class of implicit (and differentiable) probabilistic
+  models. These models do not require a tractable density and assume
+  only a program that generates samples.
+  """
+  def __init__(self, *args, **kwargs):
+    """
+    Notes
+    -----
+    ``ALI`` matches a mapping from data to latent variables and a
+    mapping from latent variables to data through a joint
+    discriminator. The encoder approximates the posterior p(z|x)
+    when the network is stochastic.
```

Comment:
Actually this is untrue. The terminology "Adversarially learned inference" is a misnomer because the distribution p( z | x) in GANs is degenerate. More accurately would be to say
```
Using a joint discriminator, ``ALI`` learns a mapping from noise to data 
as well as an inverse mapping from data to noise. The inference network 
is the inverse mapping.
```
We discuss this misconception  in [my paper](https://arxiv.org/abs/1702.08896) on deep implicit models (see the noise vs latent variables part of section 2.1).

I personally prefer the other paper which calls it "adversarial feature learning"/BiGANs; to me this description is more accurate. So I'd be for `BiGANInference`, but if you feel strongly about `ALI`, that's okay too.